src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexRunner.kt

```diff
@@ -545,6 +579,100 @@ object ManagedIndexRunner : ScheduledJobRunner,
+        }
+        return result
+    }
+
+    /**
+     * update metadata in config index, and save metadata in history after update
+     * this can be called 2 times in one job run, so need to save seqNo & primeTerm
+     */
+    private suspend fun updateManagedIndexMetaData(
+        managedIndexMetaData: ManagedIndexMetaData,
+        lastUpdateResult: UpdateMetadataResult? = null
+    ): UpdateMetadataResult {
+        val result = UpdateMetadataResult()
+
+        var metadata: ManagedIndexMetaData = managedIndexMetaData
+        if (lastUpdateResult != null) {
+            metadata = managedIndexMetaData.copy(seqNo = lastUpdateResult.seqNo, primaryTerm = lastUpdateResult.primaryTerm)
+        }
+
+        val indexRequest = managedIndexMetadataIndexRequest(metadata)
+        try {
+            updateMetaDataRetryPolicy.retry(logger) {
+                val indexResponse: IndexResponse = client.suspendUntil { index(indexRequest, it) }
+                result.savedMetadata = indexResponse.status() == RestStatus.OK ||
+                    indexResponse.status() == RestStatus.CREATED
+                result.seqNo = indexResponse.seqNo
+                result.primaryTerm = indexResponse.primaryTerm
+            }
+
+            GlobalScope.launch(Dispatchers.IO + CoroutineName("ManagedIndexMetaData-AddHistory")) {
+                ismHistory.addManagedIndexMetaDataHistory(listOf(metadata))
+            }
+        } catch (e: VersionConflictEngineException) {
+            logger.error("There was VersionConflictEngineException trying to update the metadata for " +
+                    "${managedIndexMetaData.index}. Message: ${e.message}", e)
+        } catch (e: Exception) {
+            logger.error("Failed to save ManagedIndexMetaData for [index=${managedIndexMetaData.index}]", e)
+        }
+
+        return result
+    }
+
+    data class UpdateMetadataResult(
+        var savedMetadata: Boolean = false,
+        var seqNo: Long = SequenceNumbers.UNASSIGNED_SEQ_NO,
+        var primaryTerm: Long = SequenceNumbers.UNASSIGNED_PRIMARY_TERM
+    )
+
+    /**
+     *  only if metadata from config index is null and metadata from cluster state is not null
+     *  then try to move metadata from cluster state to config index
+     *  if having metadata in config index, try to clear metadata in cluster state if we
+     *  haven't cleared it successfully at last time
+     */
+    @OpenForTesting
+    suspend fun handleClusterStateMetadata(input: ManagedIndexMetaData?, metadataFromClusterState: ManagedIndexMetaData?): ManagedIndexMetaData? {
+        var metadata: ManagedIndexMetaData? = input
+        if (metadataFromClusterState != null) {
+            if (metadata == null) {
+                // move metadata from cluster state metadata to config index
+                metadata = metadataFromClusterState
+                if (updateManagedIndexMetaData(metadata).savedMetadata) {
```

Comment:
What is behavior when this fails and you proceeded with the rest of the runner logic?

Reply:
now have seperate MetadataService to move metadata